modules/core/src/test/java/org/apache/ignite/testframework/GridTestUtils.java

```diff
@@ -164,32 +169,89 @@ public void ignite(IgniteEx ignite) {
+         *
+         * @param customMsg Intercepted {@link DiscoveryCustomMessage}.
+         */
+        public void beforeDiscovery(DiscoveryCustomMessage customMsg) {
+            // No-op.
+        }
+
+        /** {@inheritDoc} */
+        @Override public void afterDiscovery(DiscoverySpiCustomMessage msg) {
+            if (!(msg instanceof CustomMessageWrapper))
+                return;
+
+            afterDiscovery(unwrap((CustomMessageWrapper)msg));
+        }
+
+        /**
+         * Handles {@link DiscoveryCustomMessage}.
+         *
+         * @param customMsg Intercepted {@link DiscoveryCustomMessage}.
+         */
+        public void afterDiscovery(DiscoveryCustomMessage customMsg) {
+            // No-op.
+        }
+
+        /**
+         * Obtains {@link DiscoveryCustomMessage} from {@link CustomMessageWrapper}.
+         *
+         * @param wrapper Wrapper of {@link DiscoveryCustomMessage}.
+         */
+        private DiscoveryCustomMessage unwrap(CustomMessageWrapper wrapper) {
+            return U.field(wrapper, "delegate");
+        }
+    }
+
     /**
      * Injects {@link DiscoveryHook} into handling logic.
      */
     public static final class DiscoverySpiListenerWrapper implements DiscoverySpiListener {
         /** */
         private final DiscoverySpiListener delegate;
 
-        /** */
-        private final DiscoveryHook hook;
+        /** Interceptors of discovery messages. */
+        private final DiscoveryHook[] hooks;
 
         /**
          * @param delegate Delegate.
-         * @param hook Hook.
+         * @param hooks Interceptors of discovery messages.
          */
-        private DiscoverySpiListenerWrapper(DiscoverySpiListener delegate, DiscoveryHook hook) {
-            this.hook = hook;
+        private DiscoverySpiListenerWrapper(DiscoverySpiListener delegate, DiscoveryHook[] hooks) {
+            this.hooks = hooks;
             this.delegate = delegate;
         }
 
         /** {@inheritDoc} */
-        @Override public IgniteFuture<?> onDiscovery(int type, long topVer, ClusterNode node, Collection<ClusterNode> topSnapshot, @Nullable Map<Long, Collection<ClusterNode>> topHist, @Nullable DiscoverySpiCustomMessage spiCustomMsg) {
-            hook.handleDiscoveryMessage(spiCustomMsg);
+        @Override public IgniteFuture<?> onDiscovery(
+            int type,
+            long topVer,
+            ClusterNode node,
+            Collection<ClusterNode> topSnapshot,
+            @Nullable Map<Long, Collection<ClusterNode>> topHist,
+            @Nullable DiscoverySpiCustomMessage spiCustomMsg
+        ) {
+            for (DiscoveryHook hook : hooks)
+                hook.beforeDiscovery(spiCustomMsg);
 
             IgniteFuture<?> fut = delegate.onDiscovery(type, topVer, node, topSnapshot, topHist, spiCustomMsg);
 
-            hook.afterDiscovery(spiCustomMsg);
+            for (DiscoveryHook hook : hooks)
+                hook.afterDiscovery(spiCustomMsg);
```

Comment:
The `afterDiscovery` will be called not after discovery notifing listeners. `delegate.onDiscovery` is async.